packages/optimistic-oracle/src/keeper.js

```diff
@@ -0,0 +1,250 @@
+      const collateralToken = new this.web3.eth.Contract(getAbi("ExpandedERC20"), priceRequest.currency);
+      const [currentCollateralAllowance] = await Promise.all([
+        collateralToken.methods.allowance(this.account, this.ooContract.options.address).call()
+      ]);
+      if (this.toBN(currentCollateralAllowance).lt(this.toBN(MAX_UINT_VAL).div(this.toBN("2")))) {
+        const collateralApprovalTx = await collateralToken.methods
+          .approve(this.ooContract.options.address, MAX_UINT_VAL)
+          .send({
+            from: this.account,
+            gasPrice: this.gasEstimator.getCurrentFastPrice()
+          });
+        this.logger.info({
+          at: "OptimisticOracle#Keeper",
+          message: "Approved OO to transfer unlimited collateral tokens 💰",
+          currency: collateralToken.options.address,
+          collateralApprovalTx: collateralApprovalTx.transactionHash
+        });
+      }
+    }
+  }
+
+  // Create the pricefeed for a specific identifier and save it to the state, or
+  // return the saved pricefeed if already constructed.
+  async _createOrGetCachedPriceFeed(identifier) {
+    // First check for cached pricefeed for this identifier and return it if exists:
+    let priceFeed = this.priceFeedCache[identifier];
+    if (priceFeed) return priceFeed;
+
+    // No cached pricefeed found for this identifier. Create a new one.
+    // First, construct the  cofig for this identifier. We start with the `defaultPriceFeedConfig`
+    // properties and add custom properties for this specific identifier such as precision.
+    let priceFeedConfig = {
+      ...this.defaultPriceFeedConfig,
+      priceFeedDecimals: getPrecisionForIdentifier(identifier)
+    };
+    this.logger.debug({
+      at: "OptimisticOracleKeeper",
+      message: "Created pricefeed configuration for identifier",
+      defaultPriceFeedConfig: this.priceFeedConfig,
+      identifier
+    });
+
+    // Create a new pricefeed for this identifier. We might consider caching these price requests
+    // for re-use if any requests use the same identifier.
+    let newPriceFeed = await createReferencePriceFeedForEmp(
+      this.logger,
+      this.web3,
+      new Networker(this.logger),
+      () => Math.round(new Date().getTime() / 1000),
+      null, // No EMP Address needed since we're passing identifier explicitly
+      priceFeedConfig,
+      identifier
+    );
+    this.priceFeedCache[identifier] = newPriceFeed;
+    return newPriceFeed;
+  }
+
+  // Submit proposals to unproposed price requests.
+  async sendProposals() {
+    this.logger.debug({
+      at: "OptimisticOracleKeeper",
+      message: "Checking for unproposed price requsts to send proposals for"
+    });
+
+    for (let priceRequest of this.optimisticOracleClient.getUnproposedPriceRequests()) {
+      let priceFeed = await this._createOrGetCachedPriceFeed(priceRequest.identifier);
```

Comment:
nice! cool how you construct the feed.

do you want to make this a `const`?

Reply:
Yes great call